data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconChainHeadHandler.java

```diff
@@ -58,41 +55,50 @@ public BeaconChainHeadHandler(ChainStorageClient client, JsonProvider jsonProvid
+        @OpenApiResponse(status = RES_OK, content = @OpenApiContent(from = BeaconChainHead.class)),
         @OpenApiResponse(status = RES_NO_CONTENT, description = NO_CONTENT_PRE_GENESIS),
         @OpenApiResponse(status = RES_INTERNAL_ERROR)
       })
   @Override
-  public void handle(Context ctx) throws JsonProcessingException {
-    Bytes32 head_block_root = client.getBestBlockRoot();
-    if (head_block_root == null) {
-      ctx.status(SC_NO_CONTENT);
-      return;
-    }
+  public void handle(final Context ctx) throws JsonProcessingException {
+    chainDataProvider
+        .getBestBlockRoot()
+        .ifPresentOrElse(
+            block ->
+                ctx.result(
+                    chainDataProvider
+                        .getStateByBlockRoot(block)
+                        .thenApplyChecked(
+                            optionalBeaconState -> {
+                              if (optionalBeaconState.isPresent()) {
+                                final BeaconState beaconState = optionalBeaconState.get();
+                                final Checkpoint finalizedCheckpoint =
+                                    beaconState.getFinalized_checkpoint();
+                                final Checkpoint justifiedCheckpoint =
+                                    beaconState.getCurrent_justified_checkpoint();
+                                final Checkpoint previousJustifiedCheckpoint =
+                                    beaconState.getPrevious_justified_checkpoint();
 
-    // derive all other state from the head_block_root
-    BeaconState beaconState = client.getStore().getBlockState(head_block_root);
-    Checkpoint finalizedCheckpoint = beaconState.getFinalized_checkpoint();
-    Checkpoint justifiedCheckpoint = beaconState.getCurrent_justified_checkpoint();
-    Checkpoint previousJustifiedCheckpoint = beaconState.getPrevious_justified_checkpoint();
+                                final BeaconChainHead chainHeadResponse =
+                                    new BeaconChainHead(
+                                        beaconState.getSlot(),
+                                        compute_epoch_at_slot(beaconState.getSlot()),
+                                        block,
+                                        finalizedCheckpoint.getEpochSlot(),
+                                        finalizedCheckpoint.getEpoch(),
+                                        finalizedCheckpoint.getRoot(),
+                                        justifiedCheckpoint.getEpochSlot(),
+                                        justifiedCheckpoint.getEpoch(),
+                                        justifiedCheckpoint.getRoot(),
+                                        previousJustifiedCheckpoint.getEpochSlot(),
+                                        previousJustifiedCheckpoint.getEpoch(),
+                                        previousJustifiedCheckpoint.getRoot());
```

Comment:
thinking about this more... lets just move the processing to ChainDataProvider, and let ChainDataProvider create and return this object rather than do any processing at the interface...
so ChainDataProvider gives something like 
```
SafeFuture<Optional<BeaconChainHead>> getBeaconChainHead()
```
and then at this level it becomes a simple function call - which also ideally means we can write other interfaces in other standards to get the same object without fully recreating something like this...